os/kernel/sched/sched_cpuoff.c

```diff
@@ -0,0 +1,135 @@
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND,
+ * either express or implied. See the License for the specific
+ * language governing permissions and limitations under the License.
+ *
+ ****************************************************************************/
+/****************************************************************************
+ * kernel/sched/sched_cpucontroloff.c
+ *
+ * Description:
+ *   This file implements the CPU control's power off function  which handles
+ *   CPU off operations both from user applications (via CPU driver) and
+ *   from the PM idle logic.
+ *
+ ****************************************************************************/
+ /****************************************************************************
+ * Included Files
+ ****************************************************************************/
+#include <tinyara/config.h>
+#include <tinyara/arch.h>
+#include <tinyara/spinlock.h>
+#include <tinyara/irq.h>
+#include <tinyara/sched.h>
+#include <sched.h>
+#include <debug.h>
+#include <errno.h>
+#include <assert.h>
+#include <queue.h>
+
+#ifdef CONFIG_SMP
+#include <../arch/arm/src/armv7-a/smp.h>
+#include <../arch/arm/src/armv7-a/barriers.h>
+#endif
+
+#include "sched/sched.h"
+
+#ifdef CONFIG_CPU_MANAGER
+
+/****************************************************************************
+ * Pre-processor Definitions
+ ****************************************************************************/
+
+/****************************************************************************
+ * Private Types
+ ****************************************************************************/
+
+/****************************************************************************
+ * Private Data
+ ****************************************************************************/
+
+/* Spinlock for synchronizing access to CPU hotplug states */
+volatile spinlock_t g_cpuhp_lock[CONFIG_SMP_NCPUS];
+
+/* Spinlock for synchronizing access to global hotplug state */
+volatile spinlock_t g_state_transition_lock = SP_UNLOCKED;
+
+/****************************************************************************
+ * Private Functions
+ ****************************************************************************/
+
+/****************************************************************************
+ * Public Functions
+ ****************************************************************************/
+
+/****************************************************************************
+ * Name: sched_cpuoff
+ *
+ * Description:
+ *   Request to bring a CPU to power off state.
+ *
+ * Input Parameters:
+ *   cpu - The CPU index to turn of
+ *   migrate - flag to enable task migration.
+ *
+ * Returned Value:
+ *   OK (0) on success, negative value on failure.
+ *
+ ****************************************************************************/
+
+int sched_cpuoff(int cpu, bool migrate)
+{
+	if(spin_trylock(&g_cpuhp_lock[cpu])  == SP_LOCKED) {
```

Comment:
Don't you need to check the `cpu` validation?

Reply:
Oh, that's at line 103. But you use the variable here before checking. Is it ok?